channeldb/migrations.go

```diff
@@ -458,3 +459,99 @@ func paymentStatusesMigration(tx *bolt.Tx) error {
 
 	return nil
 }
+
+// migratePruneEdgeUpdateIndex is a database migration that attempts to resolve
+// some lingering bugs with regards to edge policies and their update index.
+// Stale entries within the edge update index were not being properly pruned due
+// to a miscalculation on the offset of an edge's policy last update. This
+// migration also fixes the case where the public keys within edge policies were
+// being serialized with an extra byte, causing an even greater error when
+// attempting to perform the offset calculation described earlier.
+func migratePruneEdgeUpdateIndex(tx *bolt.Tx) error {
+	// To begin the migration, we'll retrieve the update index bucket. If it
+	// does not exist, we have nothing left to do so we can simply exit.
+	edges := tx.Bucket(edgeBucket)
+	if edges == nil {
+		return nil
+	}
+	edgeUpdateIndex := edges.Bucket(edgeUpdateIndexBucket)
+	if edgeUpdateIndex == nil {
+		return nil
+	}
+
+	// Retrieve some buckets that will be needed later on. These should
+	// already exist given the assumption that the buckets above do as well.
+	edgeIndex := edges.Bucket(edgeIndexBucket)
+	if edgeIndex == nil {
+		return errors.New("edge index should exist but does not")
+	}
+	nodes := tx.Bucket(nodeBucket)
+	if nodes == nil {
+		return errors.New("node bucket should exist but does not")
+	}
+
+	log.Info("Migrating database to properly prune edge update index")
+
+	// We'll need to properly prune all the outdated entries within the edge
+	// update index. To do so, we'll gather all of the existing policies
+	// within the graph to re-populate them later on.
+	var edgeKeys [][]byte
+	err := edges.ForEach(func(edgeKey, edgePolicyBytes []byte) error {
```

Comment:
Previously there was a check to ensure that the edge key is 41 bytes, which has since been removed. `ForEach` will also return the key of any subbuckets keys and `nil` value, which includes the edge update index as it resides in `edges`.

The previous check would likely have prevented us from adding subbbucket keys to the `edgeKeys`, though instead we should add a check to skip the entry if `edgePolicyBytes` is nil. We will still require the check for `unknownPolicy`, as it is an initialized slice of length 0.

Reply:
Ah yes, not sure why I thought we were iterating over the update index bucket rather than the edges bucket. Fixed!

Reply:
It is probably okay for now, but the proper way of filtering out buckets is to check if the value is  equal to `nil` iirc. Checking the chanID length is brittle, as there is no gurantee that the bucket key is not of that length.

Reply:
Actually I think we should fix this, as the migration will be staying in the codebase.